mne/tests/test_epochs.py

```diff
@@ -61,14 +60,151 @@ def _get_data():
 clean_warning_registry()  # really clean warning stack
 
 
+def test_reject():
+    """Test epochs rejection
+    """
+    raw, events, picks = _get_data()
+    # cull the list just to contain the relevant event
+    events = events[events[:, 2] == event_id, :]
+    selection = np.arange(3)
+    drop_log = [[]] * 3 + [['MEG 2443']] * 4
+    assert_raises (TypeError, pick_types, raw)
+    picks_meg = pick_types(raw.info, meg=True, eeg=False)
+    assert_raises(TypeError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks, preload=False, reject='foo')
+    assert_raises(ValueError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks_meg, preload=False, reject=dict(eeg=1.))
+    assert_raises(KeyError, Epochs, raw, events, event_id, tmin, tmax,
+                  picks=picks, preload=False, reject=dict(foo=1.))
+
+    for preload in (True, False):
+        # no rejection
+        epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                        preload=preload)
+        assert_raises(ValueError, epochs.drop_bad_epochs, reject='foo')
+        epochs.drop_bad_epochs()
+        assert_equal(len(epochs), len(events))
+        assert_array_equal(epochs.selection, np.arange(len(events)))
+        assert_array_equal(epochs.drop_log, [[]] * 7)
+
+        # with rejection
+        epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                        reject=reject, preload=preload)
+        epochs.drop_bad_epochs()
+        assert_equal(len(epochs), len(events) - 4)
+        assert_array_equal(epochs.selection, selection)
+        assert_array_equal(epochs.drop_log, drop_log)
+
+        # rejection post-hoc
+        epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                        preload=preload)
+        epochs.drop_bad_epochs()
+        assert_equal(len(epochs), len(events))
+        epochs.drop_bad_epochs(reject)
+        assert_equal(len(epochs), len(events) - 4)
+        assert_array_equal(epochs.selection, selection)
```

Comment:
Yikes, good call! I've now added the tests, they uncovered a bug, and also caused me to simplify (I hope) the `_get_data` function.